### DISCUSSION TITLE
The memory block 'Elsa.Expressions.Models.MemoryBlockReference' does not exist for below workflow

### DISCUSSION THREAD

**Discussion by mremadevi** [elsa-workflows/elsa-core#5288] [Q&A]
Getting error The memory block 'Elsa.Expressions.Models.MemoryBlockReference' does not exist for below workflow. Please advise.

{
	"id": "19063866e115b482",
	"definitionId": "5a9c5d84332151a5",
	"name": "Workflow 2",
	"createdAt": "2024-04-26T18:03:01.402299+00:00",
	"version": 1,
	"toolVersion": "3.1.0.0",
	"variables": [],
	"inputs": [],
	"outputs": [],
	"outcomes": [],
	"customProperties": {},
	"isReadonly": false,
	"isSystem": false,
	"isLatest": true,
	"isPublished": false,
	"options": {
		"autoUpdateConsumingWorkflows": false
	},
	"root": {
		"type": "Elsa.Flowchart",
		"version": 1,
		"id": "626874ef88bcf26c",
		"nodeId": "Workflow1:626874ef88bcf26c",
		"metadata": {},
		"customProperties": {
			"source": "FlowchartJsonConverter.cs:45",
			"notFoundConnections": [],
			"canStartWorkflow": false,
			"runAsynchronously": false
		},
		"activities": [
			{
				"result": null,
				"id": "5212ab931be19aee",
				"nodeId": "Workflow1:626874ef88bcf26c:5212ab931be19aee",
				"name": "ActivityWithOutput1",
				"type": "Wave.Service.HostedServiceHandlers.ExternalEvents.WaveDrafted.ElsaWorkflow.ActivityWithOutput",
				"version": 1,
				"customProperties": {
					"canStartWorkflow": false,
					"runAsynchronously": false
				},
				"metadata": {
					"designer": {
						"position": {
							"x": 153.5,
							"y": 304
						},
						"size": {
							"width": 174.46875,
							"height": 50
						}
					}
				}
			},
			{
				"condition": {
					"typeName": "Boolean",
					"expression": {
						"type": "CSharp",
						"value": "(string)Output.LastResult == \"test\""
					}
				},
				"then": {
					"type": "Elsa.Flowchart",
					"version": 1,
					"id": "c90bc46199ca62bb",
					"nodeId": "Workflow1:626874ef88bcf26c:7d8083b550abe76d:c90bc46199ca62bb",
					"metadata": {},
					"customProperties": {
						"source": "FlowchartJsonConverter.cs:45",
						"notFoundConnections": [],
						"canStartWorkflow": false,
						"runAsynchronously": false
					},
					"activities": [
						{
							"result": null,
							"id": "3d5b9ca8dde0b9d4",
							"nodeId": "Workflow1:626874ef88bcf26c:7d8083b550abe76d:c90bc46199ca62bb:3d5b9ca8dde0b9d4",
							"name": "CommitWaveActivity1",
							"type": "Wave.Service.HostedServiceHandlers.ExternalEvents.WaveDrafted.ElsaWorkflow.CommitWaveActivity",
							"version": 1,
							"customProperties": {
								"canStartWorkflow": false,
								"runAsynchronously": false
							},
							"metadata": {
								"designer": {
									"position": {
										"x": -474.5,
										"y": -221
									},
									"size": {
										"width": 189.625,
										"height": 50
									}
								}
							}
						}
					],
					"connections": []
				},
				"else": {
					"type": "Elsa.Flowchart",
					"version": 1,
					"id": "a2307a0b4b92bb79",
					"nodeId": "Workflow1:626874ef88bcf26c:7d8083b550abe76d:a2307a0b4b92bb79",
					"metadata": {},
					"customProperties": {
						"source": "FlowchartJsonConverter.cs:45",
						"notFoundConnections": [],
						"canStartWorkflow": false,
						"runAsynchronously": false
					},
					"activities": [
						{
							"result": null,
							"id": "71b62412efbfc6ca",
							"nodeId": "Workflow1:626874ef88bcf26c:7d8083b550abe76d:a2307a0b4b92bb79:71b62412efbfc6ca",
							"name": "ErrorHandlerActivity1",
							"type": "Wave.Service.HostedServiceHandlers.ExternalEvents.WaveDrafted.ElsaWorkflow.ErrorHandlerActivity",
							"version": 1,
							"customProperties": {
								"canStartWorkflow": false,
								"runAsynchronously": false
							},
							"metadata": {
								"designer": {
									"position": {
										"x": -380.5,
										"y": -129
									},
									"size": {
										"width": 184.625,
										"height": 50
									}
								}
							}
						}
					],
					"connections": []
				},
				"result": null,
				"id": "7d8083b550abe76d",
				"nodeId": "Workflow1:626874ef88bcf26c:7d8083b550abe76d",
				"name": "If1",
				"type": "Elsa.If",
				"version": 1,
				"customProperties": {
					"canStartWorkflow": false,
					"runAsynchronously": false
				},
				"metadata": {
					"designer": {
						"position": {
							"x": 540,
							"y": 320
						},
						"size": {
							"width": 376.453125,
							"height": 120
						}
					}
				}
			}
		],
		"connections": [
			{
				"source": {
					"activity": "5212ab931be19aee",
					"port": "Done"
				},
				"target": {
					"activity": "7d8083b550abe76d",
					"port": "In"
				}
			}
		]
	}
}

**customActivity**

 public class ActivityWithOutput : CodeActivity<string>
 {
     public ActivityWithOutput( )
     { 
     }

     protected override void Execute(ActivityExecutionContext context)
     { 
         var result = "My test output";
         context.SetResult(result);
     }
 }

**Comment by glime-ai** [comment]
The error message "The memory block 'Elsa.Expressions.Models.MemoryBlockReference' does not exist" suggests that there is an issue with accessing the output of an activity within the workflow. In the provided workflow JSON, there is an "If" activity that attempts to evaluate a condition using the output of the previous activity. The condition is written as `(string)Output.LastResult == "test"`, which seems to be trying to cast the last result to a string and compare it with "test".

However, the error indicates that the workflow engine is unable to find the memory block where the output of the previous activity is supposed to be stored. This could be due to several reasons, such as the output not being set correctly, an issue with the activity implementation, or a problem with the workflow definition.

To resolve this issue, you should ensure that:

1. The `ActivityWithOutput` activity is correctly setting the result using `context.SetResult(result);`.
2. The `Output.LastResult` is being accessed correctly in the "If" activity's condition. You may need to verify that the output is being stored and is accessible at the point where the condition is evaluated.
3. The workflow definition is correctly configured to pass the output from one activity to the next.

If you continue to experience issues, you may need to review the documentation for Elsa 3.x to ensure that you are using the correct syntax and methods for setting and accessing activity outputs. Additionally, you may want to check the Elsa GitHub issues or forums for similar problems encountered by other users.

Since the context provided does not include specific documentation or examples for the exact version of Elsa being used (Elsa 3.x), I recommend checking the official Elsa documentation or reaching out to the Elsa community for further assistance.

This comment was generated by [Glime](https://glimelab.ai/?utm_source=dRbvTb7IhCSYiI&utm_medium=github).
